File: server/src/main/java/org/hkurh/doky/filestorage/FileStorageServiceImpl.java
package org.hkurh.doky.filestorage;

import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang3.RandomStringUtils;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.env.Environment;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;


@Service
public class FileStorageServiceImpl implements FileStorageService {

    private static final Log LOG = LogFactory.getLog(FileStorageServiceImpl.class);
    public static final String STORAGE_PATH_PROPERTY = "doky.filestorage.path";
    public static final String DEFAULT_STORAGE_PATH = "./mediadata";

    private Environment environment;
    private FileStorage fileStorage;

    @Override
    public String store(@NonNull MultipartFile file, @NonNull String filePath) throws IOException {
        var isFileExists = getFileStorageStrategy().checkExistence(filePath);
        if (isFileExists) {
            getFileStorageStrategy().saveFile(file, filePath);
            return filePath;
        } else {
            var extension = FilenameUtils.getExtension(file.getOriginalFilename());
            var storagePath = getStoragePath();
            var fileName = generateFileName(extension);
            var generatedFilePath = storagePath + fileName;
            getFileStorageStrategy().saveFile(file, storagePath, fileName);
            return generatedFilePath;
        }
    }

    @Override
    public Path getFile(@NonNull String filePath) throws IOException {
        return getFileStorageStrategy().getFile(filePath);
    }

    private String getStoragePath() {
        var basePath = getEnvironment().getProperty(STORAGE_PATH_PROPERTY, DEFAULT_STORAGE_PATH);
        var today = DateTime.now(DateTimeZone.getDefault());
        var fullPath = basePath + File.separator
                + today.getYear() + File.separator
                + today.getMonthOfYear() + File.separator;

        return fullPath;
    }

    private String generateFileName(String extension) {
        var randomName = RandomStringUtils.random(10, true, true);
        return randomName + "." + extension;
    }

    public Environment getEnvironment() {
        return environment;
    }

    @Autowired
    public void setEnvironment(Environment environment) {
        this.environment = environment;
    }

    public FileStorage getFileStorageStrategy() {
        return fileStorage;
    }

    @Autowired
    public void setFileStorageStrategy(FileStorage fileStorage) {
        this.fileStorage = fileStorage;
    }
}

File: server/src/main/java/org/hkurh/doky/filestorage/FileStorageServiceImpl.kt
package org.hkurh.doky.filestorage

import org.apache.commons.io.FilenameUtils
import org.apache.commons.lang3.RandomStringUtils
import org.apache.commons.logging.LogFactory
import org.joda.time.DateTime
import org.joda.time.DateTimeZone
import org.springframework.core.env.Environment
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile
import java.io.File.separator
import java.io.IOException
import java.nio.file.Path

@Service
class FileStorageServiceImpl(private val environment: Environment, private val fileStorage: FileStorage) : FileStorageService {
    @Throws(IOException::class)
    override fun store(file: MultipartFile, filePath: String?): String {
        val isFileExists = fileStorage.checkExistence(filePath)
        return if (isFileExists) {
            fileStorage.saveFile(file, filePath!!)
            filePath
        } else {
            val extension = FilenameUtils.getExtension(file.originalFilename)
            val storagePath = generateStoragePath()
            val fileName = generateFileName(extension)
            val generatedFilePath = storagePath + fileName
            fileStorage.saveFile(file, storagePath, fileName)
            generatedFilePath
        }
    }

    @Throws(IOException::class)
    override fun getFile(filePath: String): Path? {
        return fileStorage.getFile(filePath)
    }

    fun generateStoragePath(): String {
        val basePath = environment.getProperty(STORAGE_PATH_PROPERTY, DEFAULT_STORAGE_PATH)
        val today = DateTime.now(DateTimeZone.getDefault())
        return "$basePath$separator${today.year}$separator${today.monthOfYear}$separator"
    }

    private fun generateFileName(extension: String): String {
        val randomName = RandomStringUtils.random(10, true, true)
        return "$randomName.$extension"
    }

    companion object {
        private val LOG = LogFactory.getLog(FileStorageServiceImpl::class.java)
        const val STORAGE_PATH_PROPERTY = "doky.filestorage.path"
        const val DEFAULT_STORAGE_PATH = "./mediadata"
    }
}
